[contrib][linux] Fix build after introducing ASM HUF implementation

solbjorn opened the pull request:
This mostly fixes kernel build after introducing ASM HUF implementation, but also adds the ability to build the kernel with recently landed `CONFIG_WERROR` while in-kernel ZSTD library still uses the functions deprecated in 1.5.0.

Compile and run time tested on Linux 5.15-rc2 with Clang 12.

solbjorn committed:
[contrib][linux] Fix build with CONFIG_WERROR

Linux 5.15 introduces a new Kconfig option, CONFIG_WERROR, which
forces -Werror for the entire kernel.
Current in-kernel ZSTD implementation uses functions deprecated
in 1.5.0, and thus fails on -Wdeprecated-declarations.
Turn this particular error into warning to be able to build the
kernel with CONFIG_WERROR. I'm not disabling them completely to
make sure they'll be visible and [hopefully] fixed sooner or later.

Signed-off-by: Alexander Lobakin <alobakin@pm.me>

solbjorn committed:
[contrib][linux] Add huf_decompress_amd64.o target to Makefile

Commit a5f2c4552803 ("Huffman ASM") added a new ASM source file,
but it wasn't added to the kernel Makefile despite that it received
support for Huffman ASM according to the internal definitions. This
leads to undefined references, as huf_decompress.o now calls those
ASM functions.
Add it to the list of sources when building inside the kernel tree.
Kbuild can handle .S files just fine, so none additional rules
needed.

Fixes: a5f2c4552803 ("Huffman ASM")
Signed-off-by: Alexander Lobakin <alobakin@pm.me>

terrelln commented:
Thanks for the PR! The `linux.mk` changes look good. The macro changes seem right, but I have to think them through to be sure. Looks like we're missing a test case for `!DYNAMIC_BMI2` with BMI2 enabled, I'll add that.

terrelln commented:
@solbjorn I'm going to fix the macro issues in PR #2791, where I also add a test case that covers them.

Once that is merged (later today or tomorrow), can you rebase on top of the latest dev branch, and just include the config changes?

solbjorn commented:
@terrelln, sure, great!

solbjorn committed:
[contrib][linux] Add contrib/linux-kernel/linux to .gitignore

The mentioned path is being created/used by the 'import' rule for
generating source files for Linux kernel.

Signed-off-by: Alexander Lobakin <alobakin@pm.me>

solbjorn commented:
Updated.

terrelln commented:
Thanks for the PR @solbjorn!
